### PR TITLE
Improve default error

### LINKED ISSUE #34: Position in error
At the moment the parser gives errors like `(Left Could not match character '"')`, would be handy to have position information about the error.

### LINKED ISSUE #56: ParseError is `data` when it should be `newtype`
I can't see a reason why this shouldn't be using `newtype`:
```purescript
-- | The type of parsing errors.
newtype ParseError = ParseError String
```

### PR DESCRIPTION
**Pull request by JordanMartinez**
## What does this pull request do?

Fixes #56 
Fixes #34

### CODE REVIEW

**Comment by thomashoneyman** [comment]
It's really unfortunate that `ParseError` is defined as a newtype over a `String` instead of being a pair of the error and position. For example, that's what the `parsing` library does:

https://github.com/purescript-contrib/purescript-parsing/blob/f8dae0507998ef25b0d3532423733af378629913/src/Text/Parsing/Parser.purs#L36

With that type, `runParser` still produces a `ParseError`:

https://github.com/purescript-contrib/purescript-parsing/blob/f8dae0507998ef25b0d3532423733af378629913/src/Text/Parsing/Parser.purs#L70-L72

and you can recover the same error as this library with:

https://github.com/purescript-contrib/purescript-parsing/blob/f8dae0507998ef25b0d3532423733af378629913/src/Text/Parsing/Parser.purs#L38-L39

or write a `printParseError` function that pretty-prints the error and location information as you'd have both here. I know it would be a breaking change and perhaps not worth doing for that reason, but I'm more inclined to just fix it that way. I'm a little curious what you, @JordanMartinez, as well as other contributors (@hdgarrood, @garyb, @justinwoo) think of switching to the `parsing` approach.

---

That aside, this PR would already be a breaking change because of switching from `data` to `newtype`. If we wanted to try and make the smallest possible change we'd be best off keeping the error as `data` and adding the new `runParser'` (or `runParserWithPosition`) function only.

**Comment by JordanMartinez** [comment]
> That aside, this PR would already be a breaking change because of switching from data to newtype. If we wanted to try and make the smallest possible change we'd be best off keeping the error as data and adding the new runParser' (or runParserWithPosition) function only.

To clarify, how would it be a breaking change? AFAICT, the only thing changing is the runtime code, which isn't supposed to be touched by people anyway. I didn't add a `Newtype` instance. The data constructor was and still is exported, so pattern matching will still work.

**Comment by JordanMartinez** [comment]
> I'm a little curious what you think of switching to the parsing approach.

I would prefer the parsing approach. I did it this way because it ensures this code isn't breaking. It might be worth it to make a release with this change as is, and then a breaking changes release that switches completely to the parsing approach. That way, people can still benefit from other work done here (e.g. the `Monoid` instance).

**Comment by thomashoneyman** [comment]
> AFAICT, the only thing changing is the runtime code, which isn't supposed to be touched by people anyway. I didn't add a Newtype instance. The data constructor was and still is exported, so pattern matching will still work.

I'm sorry, you're correct. It's not a breaking change; any pattern-matching will continue working.

**Comment by thomashoneyman** [comment]
> That way, people can still benefit from other work done here (e.g. the Monoid instance).

That can still be released separately, as it's already merged. I think it's also fine to release the newtype change before any breaking changes. I might not want to release a `runParser'` function only to remove it again in the next version, though, as that feels a bit abrupt. But I'm not firm on that.

**Comment by JordanMartinez** [comment]
Should I submit a different PR that just does the newtype change? I included it here because it was something I noticed when working on this one.

**Comment by thomashoneyman** [comment]
Yea, perhaps do that so that this PR can focus only on the error type discussion. I'm happy to merge the newtype change.

**Comment by JordanMartinez** [comment]
I've updated this PR to remove the newtype change commit

**Comment by JordanMartinez** [comment]
While I would like to hear what others say, I would argue that we make this breaking change now rather than later for two reasons. First, most people expect `runParser` to also show where in the String an error occurred because otherwise the error message itself is relatively useless when trying to debug the parser. Second, we'll likely be making other breaking changes, so I think it would be better to do this now rather than later.

**Comment by hdgarrood** [comment]
I think changing ParseError to include a position as well as a message sounds sensible. That sounds like the kind of breaking change which would be worth doing, especially since we are about to have a round of breaking changes anyway for the upcoming 0.14 release (I’d suggest waiting until then to release this change).

**Comment by thomashoneyman** [comment]
Three in favor is pretty good. Let's go ahead and move to that representation, then, and it can be released as a breaking change along with anything else for 0.14 / anything from #59.

**Comment by JordanMartinez** [comment]
I've rebased on master, then made `runParser'` the new definition for `runParser`.

**Comment by thomashoneyman** [comment]
I was actually thinking we'd make a larger change: update the `ParseError` type to contain the position and error message, the same way as is done in the `parsing` library. Then, `runParser` would continue to have a `ParseError` in its type, except it would now have more information. We would also provide a `printParseError` (like `printJsonDecodeError`) which turns that parse error into a pretty-printed error message.

**Comment by JordanMartinez** [comment]
How should that be done when #48 is involved?

**Comment by thomashoneyman** [comment]
I'm not sure. Perhaps this change could be rolled in to #59?

**Comment by hdgarrood** [comment]
I think this change is orthogonal to #48, so the question is probably just "which one would you rather do first?"

**Comment by hdgarrood** [comment]
I would say, though, that I'd recommend keeping the scope of #59 narrow, because code points versus code units is a subtle issue which will need a bit of care taken over it.

**Comment by JordanMartinez** [comment]
If no one is against this, I'll merge this one in before merging #63.

**Comment by hdgarrood** [comment]
I thought we had reached a consensus for approach where the error type is a record containing a message and a position instead?

**Comment by JordanMartinez** [comment]
It's been so long, I couldn't remember. I'll have to read through the thread again, but most likely.

**Comment by thomashoneyman** [comment]
This message is a good summary. We're basically doing the same thing from `parsing`.

https://github.com/purescript-contrib/purescript-string-parsers/pull/57#issuecomment-691744601

**Comment by JordanMartinez** [comment]
How's this?

File: src/Text/Parsing/StringParser.purs
Comment (thomashoneyman) [inline src/Text/Parsing/StringParser.purs:33]: We could potentially create a type alias for this, too:

```suggestion
type ParseResult a = { result :: a, suffix :: PosString }

newtype Parser a = Parser (PosString -> Either ParseError (ParseResult a))
```

File: src/Text/Parsing/StringParser.purs
Comment (thomashoneyman) [inline src/Text/Parsing/StringParser.purs:35]: These documentation comments seem a bit funky to me; they keep talking about continuations, but that's not what I interpret from this function. Am I missing something?

File: src/Text/Parsing/StringParser.purs
Comment (thomashoneyman) [inline src/Text/Parsing/StringParser.purs:47]: What about having `runParser` be `forall a. Parser a -> String -> Either ParseError a`, and then having a separate `printError` function? That's a bit more flexible than the current definition (you can recover the current definition with `lmap printError`).

File: src/Text/Parsing/StringParser.purs
Comment (JordanMartinez) [reply src/Text/Parsing/StringParser.purs:33]: Why?

File: src/Text/Parsing/StringParser.purs
Comment (JordanMartinez) [reply src/Text/Parsing/StringParser.purs:35]: I was thinking the same thing.

File: src/Text/Parsing/StringParser.purs
Comment (thomashoneyman) [reply src/Text/Parsing/StringParser.purs:33]: The same record is used in several places, so this would just reduce a bit of duplication. It's not necessary, though.

File: src/Text/Parsing/StringParser.purs
Comment (JordanMartinez) [reply src/Text/Parsing/StringParser.purs]: Pushed in latest commits.

**Comment by JordanMartinez** [comment]
CI now passes.

File: src/Text/Parsing/StringParser.purs
Comment (hdgarrood) [reply src/Text/Parsing/StringParser.purs:35]: Yeah, I think this can be removed.

**Review by hdgarrood — APPROVED** [review]

**Comment by JordanMartinez** [comment]
How about the latest changes to the docs I've made?

**Review by hdgarrood — APPROVED** [review]

**Pull request merged**
